juju/juju
# Initial caasunit filestorage

## Description of change

First PR of many to add support for deploying CAAS units with storage. This one allows filesystem storage to be requested when deploying a charm. Storage attachments are tied to their unit and mot the machine. To reuse the IAAS storage infrastructure, unnecessary machine specific aspects were made generic. This require renaming structs and using name.Tag instead of MachineTag.

There's limitations, eg unit storage attachments not removed with unit etc. Needs many PRs to do the work.

We initially just store the intent to have storage. Provisioner worker etc all needs to be done.

Commit 1
Main changes - storage artefacts in state support attaching storage to (caas) units. CAAS storage created with the unit rather than when assigned to machine.

Commit 2
Testing changes. The filesystem storage test suite is used as a base suite for IAAS and CAAS models. Because volume storage not yet support for CAAS, some tests skipped or only on CAAS suite.

Commit 3
Fallout from struct renames and tag changes.

Commit 4
Initial support for showing CAAS storage info in the storage CLI.

## QA steps

Regression test IAAS storage
Deploy a caas charm with filesystem storage
juju list-storage (--filesystem) and check output
